File: RMG.Core/Music/Song.cs
using System.Collections.Generic;

namespace RMG.Core.Music
{
    public sealed class Song : IDuration
    {
        public Scale Scale { get; set; }

        public IReadOnlyList<Track> Tracks { get; set; }

        public IReadOnlyList<TimedEvent<Part>> Parts { get; set; }

        public NoteBasePattern NoteBasePattern { get; set; }

        public double Tempo { get; set; }

        public NoteBase NoteBase { get; set; }

        public IReadOnlyList<IReadOnlyList<Part>> RankedPartTemplates { get; set; }

        public IReadOnlyList<IReadOnlyList<Pattern>> RankedPatternTemplates { get; set; }

        public double Duration { get; set; }
    }
}

File: rmg-js/src/music/song.ts
import { Scale } from './scale';
import { Note } from './note';
import { Track } from './track';
import { DurationEntity } from './duration-entity';
import { Timeline } from '../core/timeline';
import { NoteBaseTimelineMap } from './note-base';
import { TrackMap } from '../composition/track-note-base-pattern-map';

export interface Song extends DurationEntity {
  tracks: Track[]
  scale: Timeline<Scale>
  noteBase: NoteBaseTimelineMap
  tempo: Timeline<number>
  notes: TrackMap<Note>
}
